AWSWindowsHelpers/public/ARN/Get-AWSWindowsHelpersARNDetail.ps1

```diff
@@ -0,0 +1,128 @@
+.INPUTS
+   arn - ARN of the AWS certificate
+.OUTPUTS
+   Returns a object with the following properties
+   Partition - The partition the resource is in
+   Service - The AWS Service
+   Region - The AWS region
+   AccountID - AWS account number
+   ResourceDetail - The resource ARN detail
+   ResourceType - The type of AWS resource
+.FUNCTIONALITY
+   Retrieves from a ARN the details of the AWS service
+#>
+Function Get-AWSWindowsHelpersARNDetail
+{
+    [CmdletBinding()]
+    Param(
+        [Parameter(Mandatory=$true)]
+        [ValidatePattern('^arn:*.:*.:*.:*.:*.')]
+        $arn      
+    )
+    $arnSplit = $arn -split ":"
+
+    $arnDetail = @{
+        partition = $arnSplit[1]
+        service = $arnSplit[2]
+        region = $arnSplit[3]
+        accountid = $arnSplit[4]
+    }
+
+    $resourceDetail = $arnSplit[5..$arnSplit.Length] -join ":"
+    $arnDetail.Add("resourcedetail",$resourceDetail)
+    $resourceARN = $resourceDetail -split "/"
+    $resourceList= @()
+
+    if($arnSplit.Length -le 6 -and $($resourceARN.length -eq 1) -or $arnDetail.service -eq 's3')
+    {   # Deal with ARN in format "arn:partition:service:region:account-id:resource"
+        if($($resourceARN.length -gt 1))
+        { # Handle S3 resources
+            $arnDetail.Add("resource",$resourceARN[0])
+        }
+        else 
+        {
+            $arnDetail.Add("resource",$arnSplit[5])    
+        }
+        Write-verbose "Adding [resource] $($arnDetail.resource)"
+    }
+    elseif($arnSplit.Length -ge 6 -and $($resourceARN.length -ge 1))
+    {   # Deal with ARN in format "arn:partition:service:region:account-id:resourcetype/resource"
+        if($resourceARN[0] -match ":")
+        {
+            $resourceType = $($resourceARN[0] -split ":")[0]
+        }
+        else 
+        {
+            $resourceType = $resourceARN[0]
+        }
+        $arnDetail.Add("resourcetype",$resourceType)
+        for($element=5;$element -lt $arnSplit.Length;$element++)
+        {   # Loop through elements split by ":"
+            $resourceARN = $arnSplit[$element] -split "/"
+            if($($resourceARN.length) -eq 1)
+            {   
+                if($arnSplit[$element+1] -notmatch "\/")
+                {   # Check if adjacent element contains resource type split by "/"
+                    $resourceType = Get-AWSWindowsHelpersARNResourceType -arnResourceValue $arnSplit[$element+1]
+                    $resourceItem = "$($arnSplit[$element])-$resourceType"
+                    $resourceList+=$arnSplit[$element]
+                    $arnDetail.Add($arnSplit[$element],$arnSplit[$element+1])
+                    Write-verbose "Adding resource [$($arnSplit[$element])] with value [$($arnSplit[$element+1])]"
+                    $element++  
+                }
+                continue
+            }
+
+            for($i = 0;$i -lt $resourceARN.length;$i++ )
```

Comment:
I'm lost, what's going on here?

Reply:
Some of the key values in the ARN split by a `/` contain additional elements which need to be iterated through. So I need to look ahead to determine if this is the case.

E.g. `arn:aws:cloudformation:us-east-1:123456789012:stack/MyProductionStack/abc9dbf0-43c2-11e3-a6e8-50fa526be49c` has three parts to the key values `stack/stackname/additionalidentifier` 

or

 `arn:aws:elasticloadbalancing:us-east-1:123456789012:loadbalancer/app/my-load-balancer/50dc6c495c0c9188` which has four parts to the key values `loadbalancer/type/load-balancer-name/load-balancer-id`